fhir-config/src/main/java/com/ibm/fhir/config/CallTimeMetric.java

```diff
@@ -0,0 +1,119 @@
+    private final String metricName;
+
+    // Use atomic values so we can be safely used across multiple threads if desired
+    private AtomicLong accumulatedTime = new AtomicLong();
+    private AtomicLong callCount = new AtomicLong();
+
+    // the number of items processed as part of this metric
+    private AtomicLong itemCount = new AtomicLong();
+
+    /**
+     * Constructor to create a metric to track the performance of a call or span of code
+     * @param fullMetricName
+     * @param metricName
+     */
+    public CallTimeMetric(String fullMetricName, String metricName) {
+        this.fullMetricName = fullMetricName;
+        this.metricName = metricName;
+    }
+
+    /**
+     * Get the full name of the metric (path + name)
+     * @return
+     */
+    public String getFullMetricName() {
+        return this.fullMetricName;
+    }
+
+    /**
+     * @return the accumulatedTime
+     */
+    public long getAccumulatedTime() {
+        return accumulatedTime.get();
+    }
+
+    /**
+     * Add nanos to the accumulated time and increment the call count
+     * @param nanos the time to add in nanoseconds
+     */
+    public void accumulateTime(long nanos) {
+        this.accumulatedTime.addAndGet(nanos);
+        this.callCount.addAndGet(1);
+    }
+
+    /**
+     * Add count to the itemCount
+     * @param count
+     */
+    public void accumulateItems(int count) {
+        this.itemCount.addAndGet(count);
+    }
+
+    /**
+     * Get the current item count
+     * @return
+     */
+    public long getItemCount() {
+        return this.itemCount.get();
+    }
+
+    /**
+     * @return the callCount
+     */
+    public long getCallCount() {
+        return callCount.get();
+    }
+
+    /**
+     * @return the metricName
+     */
+    public String getMetricName() {
+        return metricName;
+    }
+
+    /**
+     * Render the collection of metrics as a series of fine log messages
+     * @param logger
+     * @param metrics
+     */
+    public static void render(Logger logger, List<CallTimeMetric> metrics) {
+        final double NANOS = 1e9;
+        if (!logger.isLoggable(Level.FINE)) {
+            return;
+        }
+        for (CallTimeMetric m: metrics) {
+            StringBuilder msg = new StringBuilder();
+            double totalElapsed = m.getAccumulatedTime() / NANOS;
+            double averageElapsed = m.getCallCount() > 0 ? m.getAccumulatedTime() / NANOS / m.getCallCount() : Double.NaN;
+            msg.append("#METRIC")
+                .append(" ").append(String.format("%6d", m.getCallCount()))
+                .append(" ").append(String.format("%6d", m.getItemCount()))
+                .append(" ").append(String.format("%8.3f", totalElapsed))
+                .append(" ").append(String.format("%6.3f", averageElapsed))
+                .append(" ").append(m.getFullMetricName()) // at the end so the numbers line up for better readability
+                ;
+            logger.fine(msg.toString());
```

Comment:
so to enable metrics output, you just need `com.ibm.fhir.config.CallTimeMetric=fine`?

do you have a sample of what this output looks like?
does it print even when the system has been idle?

Reply:
Yes, to enable metrics I have this in my server.env:
```
TRACE_SPEC=*=info:com.ibm.fhir.MetricReport=fine
```

Example output:
```
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC     36      0    0.038  0.001 /M_JDBC_AFTER_COMMIT
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC     36      0    5.366  0.149 /M_JDBC_COMMIT
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC  30163      0  127.987  0.004 /M_JDBC_CREATE
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC  30163      0   45.243  0.001 /M_JDBC_CREATE/M_JDBC_ADD_ANY_RESOURCE
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC  30165      0   28.090  0.001 /M_JDBC_CREATE/M_JDBC_EXTRACT_SEARCH_PARAMS
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC     38 566067  127.098  3.345 /M_JDBC_FLUSH_TX_DATA
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC     40      0    0.562  0.014 /M_JDBC_FLUSH_TX_DATA/M_PARAM_PUBLISH
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC     38 566067  116.230  3.059 /M_JDBC_FLUSH_TX_DATA/M_PARAM_PUSH_BATCH
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC    548  35187   22.846  0.042 /M_JDBC_FLUSH_TX_DATA/M_PARAM_PUSH_BATCH/M_DATES
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC    465  23656    2.811  0.006 /M_JDBC_FLUSH_TX_DATA/M_PARAM_PUSH_BATCH/M_PROFILES
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC     38  39948   37.918  0.998 /M_JDBC_FLUSH_TX_DATA/M_PARAM_PUSH_BATCH/M_QUANTITIES
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC    549 207738   19.008  0.035 /M_JDBC_FLUSH_TX_DATA/M_PARAM_PUSH_BATCH/M_REFERENCES
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC     79   1382    0.577  0.007 /M_JDBC_FLUSH_TX_DATA/M_PARAM_PUSH_BATCH/M_STRINGS
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC    552 258156   31.573  0.057 /M_JDBC_FLUSH_TX_DATA/M_PARAM_PUSH_BATCH/M_TOKENS
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC     38      0    1.967  0.052 /M_JDBC_FLUSH_TX_DATA/M_PARAM_PUSH_BATCH/WHOLE_SYSTEM
[18/07/2022, 08:29:07:915 UTC] 0000006d MetricReport  1   #METRIC     40      0    0.001  0.000 /M_JDBC_FLUSH_TX_DATA/M_PARAM_RESOLVE_CANONICAL_VALUES
[18/07/2022, 08:29:07:916 UTC] 0000006d MetricReport  1   #METRIC     40      0    0.000  0.000 /M_JDBC_FLUSH_TX_DATA/M_PARAM_RESOLVE_CODE_SYSTEMS
[18/07/2022, 08:29:07:916 UTC] 0000006d MetricReport  1   #METRIC     40      0    3.304  0.083 /M_JDBC_FLUSH_TX_DATA/M_PARAM_RESOLVE_COMMON_TOKEN_VALUES
[18/07/2022, 08:29:07:916 UTC] 0000006d MetricReport  1   #METRIC     40      0    7.355  0.184 /M_JDBC_FLUSH_TX_DATA/M_PARAM_RESOLVE_LOGICAL_RESOURCES
[18/07/2022, 08:29:07:916 UTC] 0000006d MetricReport  1   #METRIC     40      0    0.000  0.000 /M_JDBC_FLUSH_TX_DATA/M_PARAM_RESOLVE_PARAMETER_NAMES
```

Currently metrics are reported and reset every 60 seconds. Metrics are only reported when there's server activity. When the server is idle, you won't see anything (unless of course there's some internal thread somewhere that happens to instantiate a MetricHandle. Only metrics that have been created within that 60 second window would be reported.